File: android/app/src/main/java/com/gastosapp/di/DatabaseModule.java
package com.gastosapp.di;

import android.content.Context;
import androidx.room.Room;
import com.gastosapp.data.local.AppDatabase;
import com.gastosapp.data.local.ExpenseDao;
import com.gastosapp.data.local.ReminderDao;
import com.gastosapp.data.local.UserDao;
import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;
import javax.inject.Singleton;

@Module
@InstallIn(SingletonComponent.class)
public class DatabaseModule {

    @Provides
    @Singleton
    public static AppDatabase provideDatabase(@ApplicationContext Context context) {
        return Room.databaseBuilder(context, AppDatabase.class, "gastos_db")
                .fallbackToDestructiveMigration()
                .build();
    }

    @Provides
    public static ExpenseDao provideExpenseDao(AppDatabase database) {
        return database.expenseDao();
    }

    @Provides
    public static ReminderDao provideReminderDao(AppDatabase database) {
        return database.reminderDao();
    }

    @Provides
    public static UserDao provideUserDao(AppDatabase database) {
        return database.userDao();
    }
}

File: app/src/main/kotlin/com/gastosapp/core/di/DatabaseModule.kt
package com.gastosapp.core.di

import android.content.Context
import androidx.room.Room
import com.gastosapp.core.data.AppDatabase
import com.gastosapp.features.expenses.data.ExpenseDao
import com.gastosapp.features.auth.data.UserDao
import com.gastosapp.features.reminders.data.ReminderDao
import com.gastosapp.features.analytics.data.BudgetDao
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object DatabaseModule {

    @Provides
    @Singleton
    fun provideDatabase(@ApplicationContext context: Context): AppDatabase {
        return Room.databaseBuilder(
            context,
            AppDatabase::class.java,
            "gastos_db"
        ).build()
    }

    @Provides
    fun provideExpenseDao(database: AppDatabase): ExpenseDao {
        return database.expenseDao()
    }

    @Provides
    fun provideReminderDao(database: AppDatabase): ReminderDao {
        return database.reminderDao()
    }

    @Provides
    fun provideUserDao(database: AppDatabase): UserDao {
        return database.userDao()
    }

    @Provides
    fun provideBudgetDao(database: AppDatabase): BudgetDao {
        return database.budgetDao()
    }
}
